### ISSUE TITLE
Requesting implementations for Hash, and TryInto<i/u/f/*>

### ISSUE THREAD

**Issue by rscarson** [mamrhein/fpdec.rs#8]
The inability to convert between rust numeric primitives and Decimal is a limitation I am running into

I am currently converting the decimal to a string and re-parsing it, which is far from ideal

An implementation of [Hash](std::hash::Hash) would also simplify things on my end

**Comment by mamrhein** [comment]
Please explain your use cases.
Converting a decimal to a signed int is quite easy: d.trunc().coefficient().
But what should be the semantic of converting a Decimal to an unsigned int?
And why do you want to convert a Decimal to a float? Mixing Decimals and floats makes it very hard to contol rounding errors!

**Comment by rscarson** [comment]
I am using this crate in the implementation of a parser having both fixed and floating point types

**Comment by rscarson** [comment]
One specific example is the implementation of the exponentiation operator - without a checked_pow for Decimal I need to downgrade the type to f64 in order to perform the calculation

Casting between types would also require the ability to convert Decimal to float

**Comment by mamrhein** [comment]
From<Decimal> for f32/f64 and TryFrom<Decimal> for primitive ints (besides usize) implemented in version 0.9.

**Comment by rscarson** [comment]
Awesome! Thanks! For hashing I can just convert to float and hash that for now

**Comment by mamrhein** [comment]
Hashing decimals is a bit tricky.
Requirements for the hash of a Decimal might be:
a) If d is equivalent to an int i, it should have the same hash value as i.
b) Two Decimals with different internal representation which compare equal, should have the same hash value.
But what about values with a fraction in general?
Converting to a float and using its hash has a sevvere drawback: two unequal Decimals can convert to the same float and thus would get the same hash value.
The best way I can think of at the moment is to convert the Decimals to its equivalent ratio of integers and hash that.

**Comment by rscarson** [comment]
I'd argue that only point b is relevant;
consider the following example:
```rust
use std::hash::Hash;
fn main() {
        let mut hasher = std::collections::hash_map::DefaultHasher::new();
        10i64.hash(&mut hasher);
        let hash_a = std::hash::Hasher::finish(&hasher);

        let mut hasher = std::collections::hash_map::DefaultHasher::new();
        10i8.hash(&mut hasher);
        let hash_b = std::hash::Hasher::finish(&hasher);

        assert_eq!(hash_a, hash_b);
}
```

The assertion will quite correctly fail:
```
thread 'main' panicked at src/main.rs:11:9:
assertion `left == right` failed
  left: 17170585774888887431
 right: 15240312484046875203
```

Equality and Hash equivalence should not match up - a `Decimal(10)` and `10i64` may have the same numeric value, but are different, and should hash different

Hashing the internal fields of Decimal should be sufficient I believe

**Comment by mamrhein** [comment]
I agree that not supporting a) is not critical, as Rust stdlib does not support it as well (as shown by your example).
But simply hashing the internal fields (i.e. derive Hash) would definitely violate b): 3.4 and 3.40 compare equal, but have different internal representations, (34, 1) and (340, 2), and thus would get different hash values.
Hashing the - reduced - equivalent ratio would solve that. So I'll go with it.

**Comment by mamrhein** [comment]
Resolved with version 0.10.0.
